versioning/worker-versions.md

```diff
@@ -0,0 +1,137 @@
+old code around until no more workflows using the old type are open.
+
+They still have the same problem with needing to tell clients to use the new
+workflow name.
+
+## Improving the situation
+
+The `patch` type APIs can't be eliminated completely. There will always be 
+situations where a "broken" workflow needs to be fixed to make progress. Such
+fixes are typically less common than just making changes to workflow logic, though.
+
+To make the more common case easier, we will provide a set of features that
+make the "versioned task queues" approach much easier to use for changes to
+workflows which break history compatability (but not interface compatability).
+
+At a high level:
+* Provide a simple API for versioning entire workers (this proposal)
+* Allow clients to point at task queue `foo`, and automatically route new workflow executions
+  to the latest version of the `foo` queue.
+* Provide good notifications and information about when old workers may be
+  decommissioned
+
+Building on this base set of features, we plan to eventually implement
+automatic safe-rollout of changes, and possibly tools like k8s operators
+which can automatically manage worker deployments.
+
+## SDK API changes
+
+The meat of this proposal is simple:
+**Add a version field to worker options in each SDK**
+
+Because version information will need to be sent to the Temporal Cluster,
+we define it as a protobuf message (which will show up in our API somewhere):
+```protobuf
+message WorkerVersions {
+    UserVersion worker_version = 1;
+}
+
+message UserVersion {
+    // The string the user passed as their version representation
+    string original = 1;
+    // Parsed from original string
+    uint32 major = 2;
+    // Parsed from original string
+    uint32 minor = 3;
+}
+```
+
+This message is pre-emptively named plurally in order to accommodate potential
+increases in versioning granularity in the future. For example, we have considered
+versioning Workflows, Activities, and Data Converters/Interceptors all separately.
+They could be added as new fields to this message if/when that happens.
+
+Each language will expose a way to (optionally) set the worker version to 
+start, supporting the possibility of other versions later. This will translate 
+into the above message to be sent to server.
+
+### TypeScript
+
+```typescript
+Worker.create({ workerVersion: '1.0', ...otherOptions })
+```
+
+### Go
+
+```go
+workerOptions := worker.Options{
+  WorkerVersion: "1.0",
+  // other options
+}
+```
+
+### Java
+
+```java
+WorkerOptions.newBuilder().setWorkerVersion("1.0").build()
+```
+
+## When and how to increment the worker version number
+
+The information here will be added as a documentation page once this feature
+has shipped.
+
+--
+
+The worker version number encompasses changes to all workflow code registered
+with the worker, as well as changes to any code which might affect the path
+of execution within workflow code.
```

Comment:
Is the implication here that a workflow version update now _must_ entail an update to deployment topology? I.e. the mapping from workfows -> workers and the balancing/tuning of worker counts in a given deployment environment probably needs to change as a downstream effect of an app developer changing this version.